### PR TITLE
feature(file buffer): Support in-memory file buffers

### PR DESCRIPTION
**Pull request by matthewerwin**
Enhance the HelloSign SDK to support file buffers instead of requiring
the use of temp-files. The support was added to include allowing the
caller to pass in the file name that is used by form.append( ) when
telling it that the Buffer should be treated as a file by giving it a
name.

Majority of any enterprise applications generate temporary files 
on a per request basis customized in some fashion. Having to
unnecessarily persist to disk just to attach the file for a signature
request is avoided by adding buffer support.

### CODE REVIEW

**Comment by npbee** [comment]
Thanks for this! I'll get this reviewed internally and report back.

**Comment by matthewerwin** [comment]
@npbee no prob. Any word on when this might be merged?

File: lib/HelloSignResource.js [hunk 1]
```diff
@@ -331,11 +331,19 @@ HelloSignResource.prototype = {
...
+                }
+                else if( data.files[i] instanceof Object &&
+                        data.files[i].hasOwnProperty('file') && data.files[i].file instanceof Buffer ) {
+                    form.append("file[" + i + "]", data.files[i].file, {filename: data.files[i].name||'signature.pdf'});
+                }
             }
-            delete data.files;
             for(var paramName in data){
-                if(data.hasOwnProperty(paramName)){
+                if(paramName != 'files' && data.hasOwnProperty(paramName)){
```
Comment (npbee) [inline lib/HelloSignResource.js:346]: Would you mind explaining why this change was required?

File: lib/HelloSignResource.js [hunk 2]
```diff
@@ -331,11 +331,19 @@ HelloSignResource.prototype = {
             // submit as form
             var form = new FormData();
             for(var i=0; i<data.files.length;i++) {
-                form.append("file[" + i + "]", fs.createReadStream(data.files[i]));
+                if( typeof data.files[i] === 'string' ) {
+                    form.append("file[" + i + "]", fs.createReadStream(data.files[i]));
+                }
+                else if( data.files[i] instanceof Buffer ) {
+                    form.append("file[" + i + "]", data.files[i], {filename: 'signature.pdf'});
```
Comment (npbee) [inline lib/HelloSignResource.js:338]: I'm a little hesitant about this default filename of `signature.pdf`.  Is there an advantage in allowing people to just specify the buffers without a filename?  If we allow buffers I'm a little more inclined to _require_ that the object form is passed with the filename included (like below).

**Review by npbee — COMMENTED** [review]
@matthewerwin, sorry for the delay and thanks for your patience!  We're doing some prioritization internally on this, but I had a chance to take a quick look and make a couple of comments.

In addition to those comments, would you mind expanding on your initial comments about the use case here?  Perhaps with a small code snippet?  I'd just like to make sure I fully understand the use case.  

I'd also love to have some tests around this, but I understand the functional tests are not really in a place where people can contribute to them easily.  I'll make an internal task to come back around to those tests.

File: lib/HelloSignResource.js (hunk 1 above)
Comment (matthewerwin) [reply lib/HelloSignResource.js:346]: Sure. The old code deleted the property "files" because it was just a string & once a file stream was opened the file name string was no longer needed. However, in the case of a buffer being passed in the "delete data.files" would delete the buffer. So, we just skip the 'files' property while serializing instead of deleting it first.

File: lib/HelloSignResource.js (hunk 2 above)
Comment (matthewerwin) [reply lib/HelloSignResource.js:338]: A file name must be provided for the buffer to work within the FormData class. So either an exception must be thrown or a default name used. I wanted to minimize impact to anyone's code & since the name is unused by HelloSign and only used for the purposes of getting the buffer support that's why I just hard coded it with a sensible name.

If for some reason this was to be extended to allow the user to specify a name it would also make sense to extract the file name from the file path (for non-buffer case) as well for uniformity. For practical purposes of getting in buffer support while maximizing backwards compatibility I chose to use a static name.

Comment (cmpaul) [reply lib/HelloSignResource.js:338]: Since it's not used, could it be something more descriptive to that effect?
 e.g., `placeholder_only_filename_not_used.pdf`?

File: lib/HelloSignResource.js [hunk 3]
```diff
@@ -331,11 +331,19 @@ HelloSignResource.prototype = {
...
+                }
+                else if( data.files[i] instanceof Object &&
+                        data.files[i].hasOwnProperty('file') && data.files[i].file instanceof Buffer ) {
+                    form.append("file[" + i + "]", data.files[i].file, {filename: data.files[i].name||'signature.pdf'});
```
Comment (cmpaul) [inline lib/HelloSignResource.js:342]: Can the default (`'signature.pdf'`) be pulled into a constant?

**Comment by cmpaul** [comment]
@matthewerwin Left a few comments -- can you add a unit test for this as well?

**Comment by matthewerwin** [comment]
@cmpaul made the initial set of changes but have been absolutely swamped this past month. Let me see if I can have someone on my team handle. Has been a lower priority for us because we just added the SDK into our build to avoid problems going forward if this doesn't get merged.

**Comment by nathanbuchar** [comment]
Closed due to inactivity, but we'll keep this PR in our back pocket should we ever decide to move forward with implementing this

**Comment by cmpaul** [comment]
Hi @nathanbuchar ! 👋  😄
